wait.go

```diff
@@ -246,36 +255,78 @@ type httpWaiter struct {
 
 // Wait waits for something to be accepting HTTP requests.
 func (w httpWaiter) Wait(ctx context.Context) error {
-	for {
-		if err := checkContextDone(ctx, w.logger, w.url); err != nil {
-			return err
-		}
+	if err := waitAndRetry(ctx, w.waitOnce); err != nil {
+		return serrors.Errorf("connecting to %q: %w", w.url, err)
+	}
 
-		res, _ := w.client.Get(w.url)
-		if res == nil {
-			continue
-		}
-		res.Body.Close()
+	return nil
+}
 
-		if res.StatusCode < http.StatusInternalServerError {
-			w.logger.DebugContext(ctx, "established connection to address",
-				"address", w.url,
-			)
-			return nil
-		}
+func (w httpWaiter) waitOnce(ctx context.Context) waitResult {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, w.url, nil)
+	if err != nil {
+		return waitResult{FatalErr: serrors.WithStack(err)}
 	}
-}
 
-// checkContextDone checks if the provided context is done, and returns
-// an error if it is.
-func checkContextDone(ctx context.Context, logger *slog.Logger, addr string) error {
-	select {
-	case <-ctx.Done():
-		logger.DebugContext(ctx, "failed to establish connection to address",
-			"address", addr,
+	res, err := w.client.Do(req)
+	if err != nil {
+		w.logger.DebugContext(ctx, "failed to connect to address",
+			"address", w.url,
+			"error", err,
 		)
-		return serrors.Errorf("timed out connecting to %q", addr)
-	default:
-		return nil
+		return waitResult{Ok: false}
+	}
+	res.Body.Close()
+
+	w.logger.DebugContext(ctx, "received status from address",
+		"address", w.url,
+		"http_status", res.StatusCode,
+	)
+	return waitResult{
+		Ok: res.StatusCode < http.StatusInternalServerError,
+	}
+}
+
+type waitResult struct {
+	// Ok returns whether the operation was successful.
+	Ok bool
+	// FatalErr indicates the retry loops should end immediately.
+	FatalErr error
+}
+
+func waitAndRetry(
+	ctx context.Context,
+	waitFunc func(context.Context) waitResult,
+) error {
+	// In an ideal world, this is configurable. Without making any breaking
+	// changes, this is probably a reasonable interval compared to the previous
+	// behavior of "make a new request as soon as the old one finished".
+	interval := 100 * time.Millisecond
+
+	var tick <-chan time.Time
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	result := make(chan waitResult, 1)
+
+	go func() { result <- waitFunc(ctx) }()
```

Comment:
Question: shouldn't this goroutine close the `result` channel when it exits?

Reply:
Nope.

Not super sure why that would be necessary, so here's some context and if I've missed your question feel free to hit me with the follow up—
1. A channel isn't a leaky resource, so you don't have to close it when you're done with it. If a channel has some data in it, it can still be garbage collected. The reason to close a channel is to tell receivers that all data that will ever be sent to it has been sent to it—here that isn't the case, since we'll be writing to the same result channel after every timer tick until the retry loop ends.
2. The execution loop runs as follows:
    1. We create a ticker that runs every 100ms, and a channel that holds up to one result.
    2. Before the first iteration of the loop, we kick off the first execution attempt in a goroutine, and set the tick channel to `nil`. Because the channel is size 1 and only one thing is attempting to write to it, as soon as the goroutine finishes execution, it can successfully write to that channel and exit. This is what prevents the goroutine from leaking whether or not anything is reading from it (and goroutines definitely can leak if orchestrated incorrectly). 
    3. We reach the select. Because `tick` is `nil`, we can ignore that path. So we're waiting either for the context to be canceled, or for the `result` channel to have a `waitResult` from the goroutine we kicked off in step (2).
    4. If the context is done first, we exit, the goroutine might still be working, but as soon as it's done (which is probably near instant because it should respect contexts), it will exit because the channel has capacity to hold the result, so that channel write is non-blocking. Then we're done.
    5. If `result` is populated first, we pull the one item out of that channel (which means the channel is now empty). Then we either exit (success/fatal error), or we set the `tick` back to `timer.C` and resume the loop. 
    6. Now that we re-enter the loop, there is nothing writing to the `result` channel at this point in time so we can ignore that code path. Instead we're now waiting between context cancellation and the next tick. If the context is canceled we exit, and the ticker is stopped and garbage collected.
    7. If the ticker goes first, we set the `tick` to `nil` so that channel starts blocking, then spin up a goroutine equivalent to the very first goroutine we kicked off. This actually resets our state all the way back to step (3)—we have a goroutine doing work, we're in the select, `tick` is `nil`, and our channel capacity is `0` with nothing in it yet.

So `result` should definitely never be closed before the `waitAndRetry` function exits, because we continually write to it over the life of the function. But also, it doesn't have to close _after_ we exit, because it's just going to be garbage collected anyway. Which means there's really no reason to close it at all.